segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/Throttler.java

```diff
@@ -0,0 +1,201 @@
+     *
+     * @param containerId Id of the Segment Container for this Throttler. Used for logging purposes only.
+     * @param calculator  A {@link ThrottlerCalculator} to be used for determining how much to throttle.
+     * @param executor    An Executor for async operations.
+     * @param metrics     Metrics for reporting.
+     */
+    Throttler(int containerId, @NonNull ThrottlerCalculator calculator, @NonNull ScheduledExecutorService executor,
+              @NonNull SegmentStoreMetrics.OperationProcessor metrics) {
+        this.throttlerCalculator = calculator;
+        this.executor = executor;
+        this.metrics = metrics;
+        this.traceObjectId = String.format("Throttler[%d]", containerId);
+        this.currentDelay = new AtomicReference<>();
+        this.closed = new AtomicBoolean(false);
+    }
+
+    //endregion
+
+    //region AutoCloseable Implementation
+
+    @Override
+    public void close() {
+        this.closed.set(true);
+    }
+
+    //endregion
+
+    //region CacheUtilizationProvider.CleanupListener Implementation
+
+    @Override
+    public void cacheCleanupComplete() {
+        val currentDelay = this.currentDelay.get();
+        if (currentDelay != null && isInterruptible(currentDelay.source)) {
+            // We were actively throttling due to a reason that is eligible for re-throttling. Terminate the current
+            // throttle cycle, which should force the throttler to re-evaluate the situation.
+            log.debug("{}: Cache Cleanup complete while actively throttling ({}).", this.traceObjectId, currentDelay);
+            currentDelay.delayFuture.completeExceptionally(new ThrottlingInterruptedException());
+        }
+    }
+
+    @Override
+    public boolean isClosed() {
+        return this.closed.get();
+    }
+
+    //endregion
+
+    //region Throttling
+
+    /**
+     * Determines whether there is an immediate need to introduce a throttling delay.
+     *
+     * @return True if throttling is required, false otherwise.
+     */
+    boolean isThrottlingRequired() {
+        return this.throttlerCalculator.isThrottlingRequired();
+    }
+
+    /**
+     * Throttles if necessary using the {@link ThrottlerCalculator} passed to this class' constructor as input.
+     *
+     * @return A CompletableFuture that is either already completed (if no throttling is required) or will complete at
+     * some time in the future based on the throttling delay value returned by {@link ThrottlerCalculator}.
+     */
+    CompletableFuture<Void> throttle() {
+        val delay = new AtomicReference<ThrottlerCalculator.DelayResult>(this.throttlerCalculator.getThrottlingDelay());
+        if (!delay.get().isMaximum()) {
+            // We are not delaying the maximum amount. We only need to do this once.
+            val existingDelay = this.currentDelay.get();
+            if (existingDelay != null && existingDelay.source == delay.get().getThrottlerName()) {
```

Comment:
I am not sure I understood this condition. Is it possible that delay has larger value than existing delay yet different source?

Reply:
That's a good point. The goal was not to mistakenly cancel the original throttle delay because of a different reason, but it is entirely possible that a delay of a different reason (i.e., Backlog vs the original Caching) indicate that the Caching delay reason has completely disappeared. This would have prevented us from (correctly) cancelling the delay in those cases.

I just relaxed this constraint.